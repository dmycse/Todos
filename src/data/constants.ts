export const todoStatus =  {
  active: 'Active',
  completed: 'Completed'
};