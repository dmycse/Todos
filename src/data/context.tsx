import { createContext, ReactNode, useContext, useReducer } from 'react';
import { reducer, TodosState } from './reducer';
import { initialState } from './initialTodos';

export const TodosContext = createContext<{
    state: TodosState, 
    dispatch: React.Dispatch<any>
  }>({
    state: initialState, 
    dispatch: () => null
  });

export const TodosProvider = ({ children }: { children: ReactNode }) => {
  
  let [state, dispatch] = useReducer(reducer, initialState);

  return (
    <TodosContext.Provider value={{state, dispatch}}>
      {children}
    </TodosContext.Provider>
  );
};

export const useTodosState = () => useContext(TodosContext);