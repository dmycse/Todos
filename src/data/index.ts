export { reducer } from './reducer';
export { initialState } from './initialTodos';

export { todoStatus } from './constants';