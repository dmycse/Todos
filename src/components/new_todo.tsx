import { KeyboardEventHandler, MouseEventHandler, useRef } from "react";
import { ACTION_TYPE, useTodosState } from "../data";

import { Box, Button, Input } from "@mui/material";
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';

export let NewTodo = () => {

  let {state, dispatch} = useTodosState();

  let inputRef = useRef<HTMLInputElement>(null);

  let addNewTodo: MouseEventHandler & KeyboardEventHandler = () => {
    if (inputRef.current!.value !== '') {
      dispatch({type: ACTION_TYPE.ADD_TODO, payload: {title: (inputRef.current?.value as string).trim()}});
      inputRef.current!.value = '';
    } 
    return;
  };


  return (
    <Box
      component='div'
      sx={{
        paddingTop: 1, 
        paddingBottom: 1, 
        flexGrow: 0, 
        display: 'flex', 
        alignItems: 'center'
      }}
    >
      <Button 
        type='button'
        variant="text"
        onClick={addNewTodo}
        sx={{
          padding: '0px',
          '& :hover': {color: '#753c3c'}
        }}
      >
        <KeyboardArrowDownIcon sx={{width: '40px', height:'40px', color: '#ccc'}} />
      </Button>
      <Input 
        type="text" 
        placeholder="What needs to be done?"
        disableUnderline
        inputProps={{style: {fontFamily: 'Inter', fontSize: '20px', fontWeight: '100', fontStyle: 'italic'}}} 
        sx={{ flexBasis: '100%'}}  
        inputRef={inputRef}
        onKeyDown={e => e.key === 'Enter' && addNewTodo(e)}
        onFocus={() => dispatch({type: ACTION_TYPE.VIEW_ALL_TODO})}
        data-testid="input"
      />
    </Box>
  );
};