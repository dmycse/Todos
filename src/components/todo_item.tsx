import { ChangeEvent } from "react";
import { Stack, Checkbox, Input, Button } from "@mui/material";
import { CheckBoxDoneIcon, CheckBoxIcon } from "../ui";
import { ACTION_TYPE, useTodosState } from "../data";
import DeleteIcon from '@mui/icons-material/Delete';

type TodoItemProps = {
  id: number;
  title: string;
  completed: boolean;
};

export let TodoItem = ({ id, title, completed }: TodoItemProps) => {

  let {state, dispatch} = useTodosState();

  return (
    <Stack 
      direction='row' 
      sx={{
        padding: 1, 
        justifyContent: 'space-between', 
        alignItems: 'center', 
        gap: 1, 
        border: '0.5px solid #ccc',
      }}>
      <Checkbox
        checked={completed}
        onChange={() => dispatch({type: ACTION_TYPE.TOGGLE_TODO, payload: {id}})}
        icon={<CheckBoxIcon />}
        checkedIcon={<CheckBoxDoneIcon />}
        inputProps={{ 'aria-label': 'controlled' }}
        data-testid="checkbox"
      />
      <Input 
        type="text"
        fullWidth
        disableUnderline
        disabled={completed}
        sx={{ 
          fontSize: '20px', 
          fontFamily: 'Inter', 
          fontWeight: '200', 
          textDecoration: completed ? 'line-through' : 'none',
        }}
        value={title}
        onChange={(e: ChangeEvent<HTMLInputElement>) => {
          dispatch({type: ACTION_TYPE.UPDATE_TODO, payload: {id, title: e.target.value}});
        }}
      />
      <Button>
        <DeleteIcon
          sx={{color: '#000', opacity: '0.2', '&:hover': {opacity: '0.6'}}} 
          onClick={() => dispatch({type: ACTION_TYPE.REMOVE_TODO, payload: {id}})}/>
      </Button>
    </Stack>
  );
};