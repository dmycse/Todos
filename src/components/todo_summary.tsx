import { Box, Button, Stack } from "@mui/material";
import { ACTION_TYPE, useTodosState, VIEW_TYPE } from "../data";
import { ButtonSummary } from "../ui/button_summary";

export let TodoSummary = () => {

  let {state, dispatch} = useTodosState();
  
  let totalTodos = state.todos.length;
  let completedTodosCount = state.todos.filter(todo => todo.completed).length;
  let activeTodosCount = totalTodos - completedTodosCount;

  let handleClearCompleted = () => {
    dispatch({type: ACTION_TYPE.CLEAR_COMPLETED});
  };

  return (
    <Stack
      component='footer' 
      direction='row' 
      sx={{
        padding: '10px',
        justifyContent: 'space-between', 
        alignItems: 'center',
        borderTop: '1px solid #ccc'
      }}>
      <Box component='span' sx={{fontFamily: 'Inter', fontSize: '14px', fontWeight: '100'}}>
        {activeTodosCount} items left
      </Box>
      <Stack direction='row' justifyContent='center' alignItems='center'>
        <ButtonSummary 
          viewType={VIEW_TYPE.VIEW_ALL} 
          actionType={ACTION_TYPE.VIEW_ALL_TODO}
          disabled={totalTodos === 0}
        >
          {VIEW_TYPE.VIEW_ALL}
        </ButtonSummary>
        <ButtonSummary 
          viewType={VIEW_TYPE.VIEW_ACTIVE} 
          actionType={ACTION_TYPE.VIEW_ACTIVE_TODO}
          disabled={totalTodos === 0 || activeTodosCount === 0}
        >
          {VIEW_TYPE.VIEW_ACTIVE}
        </ButtonSummary>
        <ButtonSummary 
          viewType={VIEW_TYPE.VIEW_COMPLETED} 
          actionType={ACTION_TYPE.VIEW_COMPLETED_TODO}
          disabled={totalTodos === 0 || completedTodosCount === 0}
        >
          {VIEW_TYPE.VIEW_COMPLETED}
        </ButtonSummary>
      </Stack>
      <Button 
        variant="text"
        disabled={totalTodos === 0 || completedTodosCount === 0}
        onClick={handleClearCompleted}
        sx={{
          fontFamily: 'Inter', 
          fontSize: '14px', 
          fontWeight: '100',
          color: '#5e5e5e',
          textTransform: 'none',
        }}
      >
        Clear {VIEW_TYPE.VIEW_COMPLETED}
      </Button>
    </Stack>
  );
};