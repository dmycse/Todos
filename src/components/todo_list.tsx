import { useTodosState } from "../data";

import { TodoItem } from "./todo_item";
import { TodoSummary } from "./todo_summary";
import { NewTodo } from "./new_todo";
import { SheetsStack } from "../ui";
import { Paper, Stack, Typography } from "@mui/material";


export let TodoList = () => {

  let {state} = useTodosState();

  let todosList = {
    'All': state.todos,
    'Active': state.todos.filter(todo => !todo.completed),
    'Completed': state.todos.filter(todo => todo.completed)
   };


  if (state.todos.length === 0) {
    return (
      <>
        <Paper 
          elevation={10} 
          sx={{
            paddingTop: 2,
            height: '100%',
            display: 'flex',
            flexDirection: 'column', 
            border: '1px solid #ccc', 
            textAlign: 'center',
            boxShadow: '0 1px 1px rgba(0,0,0,0.15) ',
          }}
        >
          <NewTodo />
          <Typography variant="h6" 
            sx={{
              paddingTop: '1rem',
              paddingBottom: '1rem',
              flexGrow: 1,
              fontFamily: 'Inter', 
              fontSize: '24px', 
              fontWeight: '100',
              borderTop: '1px solid #ccc'
              }}
            >
              No todos. Add a new todo to get started.
          </Typography>
          <TodoSummary />
        </Paper>
        <SheetsStack />
      </>
    );
  };

  return (
    <>
      <Paper 
        elevation={3} 
        sx={{ 
          flexGrow: 1, 
          border: '1px solid #ccc',
          display: 'flex',
          flexDirection: 'column',
          overflow: 'auto'  
        }}
      >
        <NewTodo />
        <Stack sx={{ flexGrow: 1, overflow: 'auto', borderTop: '1px solid #ccc'}}>
          {todosList[state.view]?.length === 0 && (
            <Typography 
              variant="h6" 
              sx={{
                padding: '1rem', 
                fontFamily: 'Inter', 
                fontSize: '24px', 
                fontWeight: '100',
                textAlign: 'center'
              }}
            >
              {`No ${state.view} todos`}
            </Typography>
          )}
          {todosList[state.view]?.map(todo => <TodoItem key={todo.id} {...todo} />)}
        </Stack>
        <TodoSummary />
      </Paper>
      <SheetsStack />
    </>
  );
};