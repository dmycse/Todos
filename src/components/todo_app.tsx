import { Stack, Typography } from "@mui/material";
import { TodoList } from "./todo_list";

export let TodoApp = () => {

  return (
    <Stack
      component={'main'}
      spacing={2} 
      sx={{
        padding: '1rem', 
        width: '100%', 
        height: '85vh', 
        display: 'flex', 
        flexDirection: 'column', 
        justifyContent: 'center', 
        alignItems: 'center', 
        background: '#faf7f7'
    }}>
      <Typography 
        variant="h1" 
        sx={{fontFamily: 'Inter', fontWeight: '100', color: '#753c3c', opacity: '0.2'}}
      >
        todos
      </Typography>
      <Stack sx={{width: '100%', height: '100%', overflow: 'auto' }}>
        <TodoList />
      </Stack>
    </Stack>
  );
};