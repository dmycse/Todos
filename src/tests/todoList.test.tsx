import { cleanup, render, screen } from "@testing-library/react";
import { TodoList } from "../components/todo_list";
import { initialState } from "../data";

describe('TodoList component', () => {

  afterEach(cleanup);

  it('TodoList should render successfully', () => {
    const { baseElement } = render(<TodoList />);
    expect(baseElement).toBeTruthy();
  });
 
  it('TodoList should render initial array of 3 TodoItem', () => {
    render(<TodoList />);

    initialState.todos.forEach(todo => {
      expect(screen.getByDisplayValue(todo.title)).toBeInTheDocument();
    });
  });

  it('TodoList snapshot', () => {
    const { baseElement } = render(<TodoList />);
    expect(baseElement).toMatchSnapshot();
  });

});