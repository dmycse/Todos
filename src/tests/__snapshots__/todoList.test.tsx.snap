// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component TodoList snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-c9vlly-MuiPaper-root"
      style="--Paper-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiBox-root css-e3s1fk"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-9eepxd-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1les0pt-MuiSvgIcon-root"
            data-testid="KeyboardArrowDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
            />
          </svg>
        </button>
        <div
          class="MuiInputBase-root MuiInput-root MuiInputBase-colorPrimary css-kb18pf-MuiInputBase-root-MuiInput-root"
          data-testid="input"
        >
          <input
            class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
            placeholder="What needs to be done?"
            style="font-family: Inter; font-size: 20px; font-weight: 100; font-style: italic;"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiStack-root css-oc97an-MuiStack-root"
      >
        <div
          class="MuiStack-root css-rv51nl-MuiStack-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-imrjgg-MuiButtonBase-root-MuiCheckbox-root"
            data-testid="checkbox"
          >
            <input
              aria-label="controlled"
              class="PrivateSwitchBase-input css-j8yymo"
              data-indeterminate="false"
              type="checkbox"
            />
            <div
              class="MuiBox-root css-mipas4"
            />
          </span>
          <div
            class="MuiInputBase-root MuiInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-nph5b-MuiInputBase-root-MuiInput-root"
          >
            <input
              class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
              type="text"
              value="Test task"
            />
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1588512-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-qzfqa9-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
        <div
          class="MuiStack-root css-rv51nl-MuiStack-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-imrjgg-MuiButtonBase-root-MuiCheckbox-root"
            data-testid="checkbox"
          >
            <input
              aria-label="controlled"
              class="PrivateSwitchBase-input css-j8yymo"
              data-indeterminate="false"
              type="checkbox"
            />
            <div
              class="MuiBox-root css-mipas4"
            />
          </span>
          <div
            class="MuiInputBase-root MuiInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-nph5b-MuiInputBase-root-MuiInput-root"
          >
            <input
              class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
              type="text"
              value="Create code"
            />
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1588512-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-qzfqa9-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
        <div
          class="MuiStack-root css-rv51nl-MuiStack-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-imrjgg-MuiButtonBase-root-MuiCheckbox-root"
            data-testid="checkbox"
          >
            <input
              aria-label="controlled"
              checked=""
              class="PrivateSwitchBase-input css-j8yymo"
              data-indeterminate="false"
              type="checkbox"
            />
            <div
              class="MuiBox-root css-s5ogrw"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fdw446-MuiSvgIcon-root"
                data-testid="DoneIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4z"
                />
              </svg>
            </div>
          </span>
          <div
            class="MuiInputBase-root MuiInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-fullWidth css-aw0522-MuiInputBase-root-MuiInput-root"
          >
            <input
              class="MuiInputBase-input MuiInput-input Mui-disabled css-1yrc8ca-MuiInputBase-input-MuiInput-input"
              disabled=""
              type="text"
              value="Testing code"
            />
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1588512-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-qzfqa9-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
      </div>
      <footer
        class="MuiStack-root css-u3o20w-MuiStack-root"
      >
        <span
          class="MuiBox-root css-43cqvf"
        >
          2
           items left
        </span>
        <div
          class="MuiStack-root css-7owdng-MuiStack-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-a6of9r-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            All
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-104zis2-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Active
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-104zis2-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Completed
          </button>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-qvkudh-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Clear 
          Completed
        </button>
      </footer>
    </div>
    <div
      class="MuiBox-root css-1lio3o0"
    />
    <div
      class="MuiBox-root css-s08ci9"
    />
  </div>
</body>
`;
